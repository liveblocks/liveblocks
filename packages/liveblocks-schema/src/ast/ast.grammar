#
# Below you'll find the full AST specification. This is a simplistic
# specification format (a DSL) for specifying which AST node hierarchies are
# valid.
#
# This is used to generate the AST TypeScript module.
#

@Comment:
    # | BlockComment
    | LineComment


@Definition:
    | ObjectTypeDef


# @Literal:
#     | FloatLiteral
#     | IntLiteral
#     | StringLiteral


@TypeExpr:
    # | @Literal
    | ObjectLiteralExpr
    | TypeRef


Document:
    definitions  @Definition+
    comments     @Comment*?


FieldDef:
    name      Identifier
    optional  boolean
    type      @TypeExpr


Identifier:
    name  string


LineComment:
    text  string


# e.g. { cx: Number; cy: Number; r: Float }
ObjectLiteralExpr:
    fields  FieldDef*


# e.g. type Circle { cx: Number; cy: Number; r: Float }
ObjectTypeDef:
    name  TypeName
    obj   ObjectLiteralExpr


# StringLiteral:
#     value     string
#     rawValue  string


TypeName:
    name  string


TypeRef:
    name  TypeName
    args  @TypeExpr*
