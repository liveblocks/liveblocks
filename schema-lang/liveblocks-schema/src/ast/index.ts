export * from "./generated-ast.js";
