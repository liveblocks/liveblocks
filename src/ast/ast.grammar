#
# Below you'll find the full AST specification. This is a simplistic
# specification format (a DSL) for specifying which AST node hierarchies are
# valid.
#
# This is used to generate the AST TypeScript module.
#

@Comment:
    # | BlockComment
    | LineComment


@Definition:
    | ObjectTypeDef


@Literal:
    # | FloatLiteral
    # | IntLiteral
    | StringLiteral


@TypeExpr:
    | @Literal
    | ObjectLiteralExpr
    | @TypeRef


@TypeRef:
  | InstantiatedType
  | TypeName


Document:
    definitions  @Definition*
    comments     @Comment*


FieldDef:
    name      Identifier
    optional  boolean
    type      @TypeExpr


Identifier:
    name  string


# e.g. LiveObject<{ cx: Number; cy: Number; r: Float }>
# e.g. LiveList<Int>
InstantiatedType:
    name  TypeName  # must be "LiveObject", "LiveList", or "LiveMap"
    args  @TypeExpr+


LineComment:
    text  string


# e.g. { cx: Number; cy: Number; r: Float }
ObjectLiteralExpr:
    fields  FieldDef*


# e.g. type Circle { cx: Number; cy: Number; r: Float }
ObjectTypeDef:
    name  TypeName
    fields  FieldDef*


StringLiteral:
    value     string
    rawValue  string


TypeName:
    name  string
