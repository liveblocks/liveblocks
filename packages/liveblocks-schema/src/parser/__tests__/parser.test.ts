import * as fc from "fast-check";

import * as ast from "../../ast";
import { parseDocument as parse } from "..";
import { expectDocument, expectLegacyDocument } from "./helpers";

describe("syntactic parser", () => {
  it("fails on all non-valid inputs", () => {
    fc.assert(
      fc.property(
        fc.anything().filter((value) => typeof value !== "string"),
        fc.anything(),

        (arg1, arg2) => {
          expect(() =>
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (parse as any)(arg1, arg2)
          ).toThrow();
        }
      )
    );
  });

  it("rejects all non-valid documents", () => {
    fc.assert(
      fc.property(
        // NOTE: This test assumes that fast-check will not magically generate
        // a syntactically valid schema text. There is too much entropy for
        // that to happen :)
        fc.string(),

        (input) => {
          expect(() => parse(input as string)).toThrow();
        }
      )
    );
  });

  it("minimal document (legacy parse mode)", () => {
    expectLegacyDocument(
      `
      type Foo {
        cx: Float
        cy: Float
        r: Float
      }

      type Foo {          // Double type defs are syntactically valid
        version: Int
        version: Int      // Double field defs are syntactically valid
        mycircle?: LiveObject<Bar>
        //                    ^^^ Will parse the syntax, even though semantically incorrect
        someField: _undefinedThing_
        list: LiveList<Bar>
        //             ^^^ Will parse the syntax, even though semantically incorrect
        map: LiveMap<String, LiveList<LiveMap<Bar[], Qux>[]>>
        //                                    ^^^^^ Will parse the syntax, even though semantically incorrect
      }

      type abc {}         // Lowercased type names are syntactically valid
      `,

      ast.document([
        ast.objectTypeDefinition(
          ast.typeName("Foo"),
          [
            ast.fieldDef(ast.identifier("cx"), false, ast.numberType()),
            ast.fieldDef(ast.identifier("cy"), false, ast.numberType()),
            ast.fieldDef(ast.identifier("r"), false, ast.numberType()),
          ],
          null,
          false /* always false during the parsing phase */
        ),

        ast.objectTypeDefinition(
          ast.typeName("Foo"),
          [
            ast.fieldDef(ast.identifier("version"), false, ast.numberType()),
            ast.fieldDef(ast.identifier("version"), false, ast.numberType()),
            ast.fieldDef(
              ast.identifier("mycircle"),
              true,
              ast.typeRef(ast.typeName("Bar"), true)
            ),
            ast.fieldDef(
              ast.identifier("someField"),
              false,
              ast.typeRef(ast.typeName("_undefinedThing_"), false)
            ),

            ast.fieldDef(
              ast.identifier("list"),
              false,
              ast.liveListType(ast.typeRef(ast.typeName("Bar"), false))
            ),

            ast.fieldDef(
              ast.identifier("map"),
              false,
              ast.liveMapType(
                ast.stringType(),
                ast.liveListType(
                  ast.arrayType(
                    ast.liveMapType(
                      ast.arrayType(ast.typeRef(ast.typeName("Bar"), false)),
                      ast.typeRef(ast.typeName("Qux"), false)
                    )
                  )
                )
              )
            ),
          ],
          null,
          false /* always false during the parsing phase */
        ),

        ast.objectTypeDefinition(
          ast.typeName("abc"),
          [],
          null,
          false /* always false during the parsing phase */
        ),
      ])
    );
  });

  it("minimal document (normal parse mode)", () => {
    expectDocument(
      `
      type Foo {
        cx: number
        cy: number
        r: number
      }

      type Foo {          // Double type defs are syntactically valid
        version: number
        version: number      // Double field defs are syntactically valid
        mycircle?: LiveObject<Bar>
        //                    ^^^ Will parse the syntax, even though semantically incorrect
        someField: _undefinedThing_
        list: LiveList<Bar>
        //             ^^^ Will parse the syntax, even though semantically incorrect
        map: LiveMap<string, LiveList<LiveMap<Bar[], Qux>[]>>
        //                                    ^^^^^ Will parse the syntax, even though semantically incorrect
      }

      type abc {}         // Lowercased type names are syntactically valid
      `,

      ast.document([
        ast.objectTypeDefinition(
          ast.typeName("Foo"),
          [
            ast.fieldDef(ast.identifier("cx"), false, ast.numberType()),
            ast.fieldDef(ast.identifier("cy"), false, ast.numberType()),
            ast.fieldDef(ast.identifier("r"), false, ast.numberType()),
          ],
          null,
          false /* always false during the parsing phase */
        ),

        ast.objectTypeDefinition(
          ast.typeName("Foo"),
          [
            ast.fieldDef(ast.identifier("version"), false, ast.numberType()),
            ast.fieldDef(ast.identifier("version"), false, ast.numberType()),
            ast.fieldDef(
              ast.identifier("mycircle"),
              true,
              ast.typeRef(ast.typeName("Bar"), true)
            ),
            ast.fieldDef(
              ast.identifier("someField"),
              false,
              ast.typeRef(ast.typeName("_undefinedThing_"), false)
            ),

            ast.fieldDef(
              ast.identifier("list"),
              false,
              ast.liveListType(ast.typeRef(ast.typeName("Bar"), false))
            ),

            ast.fieldDef(
              ast.identifier("map"),
              false,
              ast.liveMapType(
                ast.stringType(),
                ast.liveListType(
                  ast.arrayType(
                    ast.liveMapType(
                      ast.arrayType(ast.typeRef(ast.typeName("Bar"), false)),
                      ast.typeRef(ast.typeName("Qux"), false)
                    )
                  )
                )
              )
            ),
          ],
          null,
          false /* always false during the parsing phase */
        ),

        ast.objectTypeDefinition(
          ast.typeName("abc"),
          [],
          null,
          false /* always false during the parsing phase */
        ),
      ])
    );
  });

  it("large document (snapshot test)", () => {
    expect(
      parse(
        `
      # Comment
      // Another comment

      type Color { r: number, g: number, b: number }

      type Circle {
        cx: number
        cy: number
        r: number
        fill?: Color[]
        stroke?: Color
        third?: { r: number, g: number, b: number }[]
      }

      type Foo {}
      type Foo {          // Double type defs are syntactically valid
        version: number
        version: number      // Double field defs are syntactically valid
        mycircle?: LiveObject<Bar>
        //                    ^^^ Will parse the syntax, even though semantically incorrect
        someField: _undefinedThing_
        list: LiveList<LiveMap<number[], LiveList<Bar[][]>>>
        //                     ^^^^^^^^ Will parse, but is invalid
      }

      type abc {}         // Lowercased type names are syntactically valid
      `
      )
    ).toMatchSnapshot();
  });

  it("large legacy document (snapshot test)", () => {
    expect(
      parse(
        `
      # Comment
      // Another comment

      type Color { r: Int, g: Int, b: Int }

      type Circle {
        cx: Float
        cy: Float
        r: Float
        fill?: Color[]
        stroke?: Color
        third?: { r: Int, g: Int, b: Int }[]
      }

      type Foo {}
      type Foo {          // Double type defs are syntactically valid
        version: Int
        version: Int      // Double field defs are syntactically valid
        mycircle?: LiveObject<Bar>
        //                    ^^^ Will parse the syntax, even though semantically incorrect
        someField: _undefinedThing_
        list: LiveList<LiveMap<Int[], LiveList<Bar[][]>>>
        //                     ^^^^^ Will parse, but is invalid
      }

      type abc {}         // Lowercased type names are syntactically valid
      `
      )
    ).toMatchSnapshot();
  });
});
