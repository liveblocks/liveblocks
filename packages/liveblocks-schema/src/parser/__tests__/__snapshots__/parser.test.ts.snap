// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syntactic parser large document (snapshot test) 1`] = `
{
  "_kind": "Document",
  "definitions": [
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "r",
            "range": [
              62,
              63,
            ],
          },
          "optional": false,
          "range": [
            62,
            71,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              65,
              71,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "g",
            "range": [
              73,
              74,
            ],
          },
          "optional": false,
          "range": [
            73,
            82,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              76,
              82,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "b",
            "range": [
              84,
              85,
            ],
          },
          "optional": false,
          "range": [
            84,
            93,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              87,
              93,
            ],
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Color",
        "range": [
          54,
          59,
        ],
      },
      "range": [
        49,
        95,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "cx",
            "range": [
              125,
              127,
            ],
          },
          "optional": false,
          "range": [
            125,
            135,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              129,
              135,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "cy",
            "range": [
              144,
              146,
            ],
          },
          "optional": false,
          "range": [
            144,
            154,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              148,
              154,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "r",
            "range": [
              163,
              164,
            ],
          },
          "optional": false,
          "range": [
            163,
            172,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              166,
              172,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "fill",
            "range": [
              181,
              185,
            ],
          },
          "optional": true,
          "range": [
            181,
            195,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "ArrayType",
            "ofType": {
              "_kind": "TypeRef",
              "asLiveObject": false,
              "range": [
                188,
                193,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "Color",
                "range": [
                  188,
                  193,
                ],
              },
            },
            "range": [
              188,
              195,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "stroke",
            "range": [
              204,
              210,
            ],
          },
          "optional": true,
          "range": [
            204,
            218,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              213,
              218,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Color",
              "range": [
                213,
                218,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "third",
            "range": [
              227,
              232,
            ],
          },
          "optional": true,
          "range": [
            227,
            272,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "ArrayType",
            "ofType": {
              "_kind": "ObjectLiteralType",
              "fields": [
                {
                  "_kind": "FieldDef",
                  "leadingComment": null,
                  "name": {
                    "_kind": "Identifier",
                    "name": "r",
                    "range": [
                      237,
                      238,
                    ],
                  },
                  "optional": false,
                  "range": [
                    237,
                    246,
                  ],
                  "trailingComment": null,
                  "type": {
                    "_kind": "NumberType",
                    "range": [
                      240,
                      246,
                    ],
                  },
                },
                {
                  "_kind": "FieldDef",
                  "leadingComment": null,
                  "name": {
                    "_kind": "Identifier",
                    "name": "g",
                    "range": [
                      248,
                      249,
                    ],
                  },
                  "optional": false,
                  "range": [
                    248,
                    257,
                  ],
                  "trailingComment": null,
                  "type": {
                    "_kind": "NumberType",
                    "range": [
                      251,
                      257,
                    ],
                  },
                },
                {
                  "_kind": "FieldDef",
                  "leadingComment": null,
                  "name": {
                    "_kind": "Identifier",
                    "name": "b",
                    "range": [
                      259,
                      260,
                    ],
                  },
                  "optional": false,
                  "range": [
                    259,
                    268,
                  ],
                  "trailingComment": null,
                  "type": {
                    "_kind": "NumberType",
                    "range": [
                      262,
                      268,
                    ],
                  },
                },
              ],
              "range": [
                235,
                270,
              ],
            },
            "range": [
              235,
              272,
            ],
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Circle",
        "range": [
          108,
          114,
        ],
      },
      "range": [
        103,
        280,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Foo",
        "range": [
          293,
          296,
        ],
      },
      "range": [
        288,
        299,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "version",
            "range": [
              378,
              385,
            ],
          },
          "optional": false,
          "range": [
            378,
            393,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              387,
              393,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "version",
            "range": [
              402,
              409,
            ],
          },
          "optional": false,
          "range": [
            402,
            467,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "NumberType",
            "range": [
              411,
              467,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "mycircle",
            "range": [
              476,
              484,
            ],
          },
          "optional": true,
          "range": [
            476,
            502,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": true,
            "range": [
              487,
              502,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Bar",
              "range": [
                498,
                501,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "someField",
            "range": [
              603,
              612,
            ],
          },
          "optional": false,
          "range": [
            603,
            630,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              614,
              630,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "_undefinedThing_",
              "range": [
                614,
                630,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "list",
            "range": [
              639,
              643,
            ],
          },
          "optional": false,
          "range": [
            639,
            691,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "LiveListType",
            "ofType": {
              "_kind": "LiveMapType",
              "keyType": {
                "_kind": "ArrayType",
                "ofType": {
                  "_kind": "NumberType",
                  "range": [
                    662,
                    668,
                  ],
                },
                "range": [
                  662,
                  670,
                ],
              },
              "range": [
                654,
                690,
              ],
              "valueType": {
                "_kind": "LiveListType",
                "ofType": {
                  "_kind": "ArrayType",
                  "ofType": {
                    "_kind": "ArrayType",
                    "ofType": {
                      "_kind": "TypeRef",
                      "asLiveObject": false,
                      "range": [
                        681,
                        684,
                      ],
                      "ref": {
                        "_kind": "TypeName",
                        "name": "Bar",
                        "range": [
                          681,
                          684,
                        ],
                      },
                    },
                    "range": [
                      681,
                      686,
                    ],
                  },
                  "range": [
                    681,
                    688,
                  ],
                },
                "range": [
                  672,
                  689,
                ],
              },
            },
            "range": [
              645,
              691,
            ],
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Foo",
        "range": [
          311,
          314,
        ],
      },
      "range": [
        306,
        766,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "abc",
        "range": [
          779,
          782,
        ],
      },
      "range": [
        774,
        842,
      ],
    },
  ],
  "range": [
    7,
    842,
  ],
}
`;

exports[`syntactic parser large legacy document (snapshot test) 1`] = `
{
  "_kind": "Document",
  "definitions": [
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "r",
            "range": [
              62,
              63,
            ],
          },
          "optional": false,
          "range": [
            62,
            68,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              65,
              68,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Int",
              "range": [
                65,
                68,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "g",
            "range": [
              70,
              71,
            ],
          },
          "optional": false,
          "range": [
            70,
            76,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              73,
              76,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Int",
              "range": [
                73,
                76,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "b",
            "range": [
              78,
              79,
            ],
          },
          "optional": false,
          "range": [
            78,
            84,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              81,
              84,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Int",
              "range": [
                81,
                84,
              ],
            },
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Color",
        "range": [
          54,
          59,
        ],
      },
      "range": [
        49,
        86,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "cx",
            "range": [
              116,
              118,
            ],
          },
          "optional": false,
          "range": [
            116,
            125,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              120,
              125,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Float",
              "range": [
                120,
                125,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "cy",
            "range": [
              134,
              136,
            ],
          },
          "optional": false,
          "range": [
            134,
            143,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              138,
              143,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Float",
              "range": [
                138,
                143,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "r",
            "range": [
              152,
              153,
            ],
          },
          "optional": false,
          "range": [
            152,
            160,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              155,
              160,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Float",
              "range": [
                155,
                160,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "fill",
            "range": [
              169,
              173,
            ],
          },
          "optional": true,
          "range": [
            169,
            183,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "ArrayType",
            "ofType": {
              "_kind": "TypeRef",
              "asLiveObject": false,
              "range": [
                176,
                181,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "Color",
                "range": [
                  176,
                  181,
                ],
              },
            },
            "range": [
              176,
              183,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "stroke",
            "range": [
              192,
              198,
            ],
          },
          "optional": true,
          "range": [
            192,
            206,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              201,
              206,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Color",
              "range": [
                201,
                206,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "third",
            "range": [
              215,
              220,
            ],
          },
          "optional": true,
          "range": [
            215,
            251,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "ArrayType",
            "ofType": {
              "_kind": "ObjectLiteralType",
              "fields": [
                {
                  "_kind": "FieldDef",
                  "leadingComment": null,
                  "name": {
                    "_kind": "Identifier",
                    "name": "r",
                    "range": [
                      225,
                      226,
                    ],
                  },
                  "optional": false,
                  "range": [
                    225,
                    231,
                  ],
                  "trailingComment": null,
                  "type": {
                    "_kind": "TypeRef",
                    "asLiveObject": false,
                    "range": [
                      228,
                      231,
                    ],
                    "ref": {
                      "_kind": "TypeName",
                      "name": "Int",
                      "range": [
                        228,
                        231,
                      ],
                    },
                  },
                },
                {
                  "_kind": "FieldDef",
                  "leadingComment": null,
                  "name": {
                    "_kind": "Identifier",
                    "name": "g",
                    "range": [
                      233,
                      234,
                    ],
                  },
                  "optional": false,
                  "range": [
                    233,
                    239,
                  ],
                  "trailingComment": null,
                  "type": {
                    "_kind": "TypeRef",
                    "asLiveObject": false,
                    "range": [
                      236,
                      239,
                    ],
                    "ref": {
                      "_kind": "TypeName",
                      "name": "Int",
                      "range": [
                        236,
                        239,
                      ],
                    },
                  },
                },
                {
                  "_kind": "FieldDef",
                  "leadingComment": null,
                  "name": {
                    "_kind": "Identifier",
                    "name": "b",
                    "range": [
                      241,
                      242,
                    ],
                  },
                  "optional": false,
                  "range": [
                    241,
                    247,
                  ],
                  "trailingComment": null,
                  "type": {
                    "_kind": "TypeRef",
                    "asLiveObject": false,
                    "range": [
                      244,
                      247,
                    ],
                    "ref": {
                      "_kind": "TypeName",
                      "name": "Int",
                      "range": [
                        244,
                        247,
                      ],
                    },
                  },
                },
              ],
              "range": [
                223,
                249,
              ],
            },
            "range": [
              223,
              251,
            ],
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Circle",
        "range": [
          99,
          105,
        ],
      },
      "range": [
        94,
        259,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Foo",
        "range": [
          272,
          275,
        ],
      },
      "range": [
        267,
        278,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "version",
            "range": [
              357,
              364,
            ],
          },
          "optional": false,
          "range": [
            357,
            369,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              366,
              369,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Int",
              "range": [
                366,
                369,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "version",
            "range": [
              378,
              385,
            ],
          },
          "optional": false,
          "range": [
            378,
            440,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              387,
              440,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Int",
              "range": [
                387,
                440,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "mycircle",
            "range": [
              449,
              457,
            ],
          },
          "optional": true,
          "range": [
            449,
            475,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": true,
            "range": [
              460,
              475,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Bar",
              "range": [
                471,
                474,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "someField",
            "range": [
              576,
              585,
            ],
          },
          "optional": false,
          "range": [
            576,
            603,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              587,
              603,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "_undefinedThing_",
              "range": [
                587,
                603,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "list",
            "range": [
              612,
              616,
            ],
          },
          "optional": false,
          "range": [
            612,
            661,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "LiveListType",
            "ofType": {
              "_kind": "LiveMapType",
              "keyType": {
                "_kind": "ArrayType",
                "ofType": {
                  "_kind": "TypeRef",
                  "asLiveObject": false,
                  "range": [
                    635,
                    638,
                  ],
                  "ref": {
                    "_kind": "TypeName",
                    "name": "Int",
                    "range": [
                      635,
                      638,
                    ],
                  },
                },
                "range": [
                  635,
                  640,
                ],
              },
              "range": [
                627,
                660,
              ],
              "valueType": {
                "_kind": "LiveListType",
                "ofType": {
                  "_kind": "ArrayType",
                  "ofType": {
                    "_kind": "ArrayType",
                    "ofType": {
                      "_kind": "TypeRef",
                      "asLiveObject": false,
                      "range": [
                        651,
                        654,
                      ],
                      "ref": {
                        "_kind": "TypeName",
                        "name": "Bar",
                        "range": [
                          651,
                          654,
                        ],
                      },
                    },
                    "range": [
                      651,
                      656,
                    ],
                  },
                  "range": [
                    651,
                    658,
                  ],
                },
                "range": [
                  642,
                  659,
                ],
              },
            },
            "range": [
              618,
              661,
            ],
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Foo",
        "range": [
          290,
          293,
        ],
      },
      "range": [
        285,
        733,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "abc",
        "range": [
          746,
          749,
        ],
      },
      "range": [
        741,
        809,
      ],
    },
  ],
  "range": [
    7,
    809,
  ],
}
`;
